Animation (#3061)
Codecademy/docs: content/swiftui/concepts/animation/animation.md

These components are essential to build animation implicitly or explicitly on a view. In the next example, implicit animation will be used:

```swift
struct ImplicitAnimation: View {
    @State var isanimation = false // @State variable that changes to activate the animation.
    
    var body: some View {
        Button {
            isanimation.toggle()
        } label: {
            Text( isanimation ? "Hello, World!" : "press me") // Ternary used to change the view based on the value of the isanimation.
                .foregroundColor(.white)
                .padding(.all)
                .background( isanimation ? Color.black : Color.red) // Ternary operator used to change the view based onn the value of the isanimation.
                .animation(.default, value: isanimation) // This is the implicit animation. This gets activated when isanimation changes in value.
        }
    }
}
```

This is the result of the code above:

![This is a gif that shows the effect of Implicit animation on the view.](https://raw.githubusercontent.com/Codecademy/docs/main/media/Implicit-Animation-SwiftUI.gif)

> In this simple animation where the background and the text of the button changes depending on the value of the `isanimation` variable. When the button is pressed, `isanimation` changes, causing the implicit animation to change the text and background of the button. As the name implies, SwiftUI creates and manages implicit animations whenever specific situations or user interface state changes take place, which means developers don't have full control over the animations. With explicit animation, developers have full control over them.

Reviewer: [Proofreading] The first half of the sentence needs clarity. Consider using the suggestion I placed below: 
```suggestion
In this simple animation, the background and the text of the button change depending on the value of the `isanimation` variable. When the button is pressed, `isanimation` changes, causing the implicit animation to change the text and background of the button. As the name implies, SwiftUI creates and manages implicit animations whenever specific situations or user interface state changes take place, which means developers don't have full control over the animations. With explicit animation, developers have full control over them.
```

## Explicit Animation using `withAnimation()`

To be able to make explicit animation, use the same three steps used in the implicit animation. This example uses explicit animation to change the position of a button:

```swift

> struct ExplicitAnimatio: View {

Author: Sorry for that my keyboard is broken

---
Title: 'Animation'

> Description: 'In response to state changes, produce seamless visual updates.'

Reviewer: [Proofreading] Developers might wonder if this describes animations. Consider using the example below to help you revise it: 
```suggestion
Description: 'In Swift, animations are a response to state changes and produce seamless visual updates.'
```

Subjects:
  - 'iOS'
  - 'Mobile Development'
Tags:
  - 'SwiftUI'
  - 'SwiftUI ViewModifiers'
  - 'iOS'
CatalogContent:
  - 'learn-iOS-course'
  - 'path/iOS-path'
---

> **Animation** is the ability to add smooth view changes to your view by changing the state of the view. There is two types of animation in swiftUI: Implicit animation and Explicit animation both of them will be explained in the sections below. it's very important to add animation to the user experience of an app, because it makes the app more interactive with the user and enhance the user experience of the app. And also Animating custom views can be done by conforming the view to the `Animatable` protocol, and telling SwiftUI about the value you want to animate.

Reviewer: [Proofreading] While mentioning the two types of animation is helpful, it would be best to just introduce them in the other sections and mostly focus on defining animations in this section.   
```suggestion
**Animation** is the ability to add smooth view changes to your view by changing the state of the view.  Adding animations to the app makes it more interactive, enhancing user experience.
```

- A variable that changes to activate the animation. it must be `@State` variable to be able to reflect the changes to the view. 
- Then after that we need to use a `Ternary ? operator`.
- Then add the animation to the view.

  This steps are essential to build animation implicitly or explicitly to your view. In the next example Implicit annimation will be used :

```Swift
struct ImplicitAnimation: View {
    @State var isanimation = false // 1- @State variable that changes to activate the animation.
    
    var body: some View {
        Button {
            isanimation.toggle()
        } label: {
            Text( isanimation ? "Hello, World!" : "press me") // 2- Ternary Operator used to change the view based on the value of the isanimation.
                .foregroundColor(.white)
                .padding(.all)
                .background( isanimation ? Color.black : Color.red) // 2- Ternary Operator used to change the view based onn the value of the isanimation.
                .animation(.default, value: isanimation) // 3- This is the Implicit animation. this gets activated when the isanimation changes.
        }

    }
}
  ```

> > **Note:** `.animation` is a view modifier, which means that views That have this view applied to will be the only views to be animated.

Reviewer: [Proofreading] Mentioning the context in which `animation` is used would make the explanation clearer. 
```suggestion
> **Note:** `.animation` in this code snippet, is a view modifier, which means that views That have this view applied to will be the only views to be animated.
```

Author: this is a general rule not only about this code snip

This is the result of the code above:

![This is a gif that shows the effect of Implicit animation on the view.](https://raw.githubusercontent.com/Codecademy/docs/main/media/Implicit-Animation-SwiftUI.gif)

> This is a very simple animation, where the background and the text of the button changes depending on the value of the `isanimation` variable, when we press the button the `isanimation` changes which active the implicit animation which leads to changing the text and the backgroung of the button in a very smooth way (try to remove animation to see the difference). As the name implies, SwiftUI creates and manages implicit animations whenever specific situations or user interface state changes take place, which means developers doesn't have full control over them and this takes us to the second type of animation Explicit animation where developers have full control over them.

Reviewer: [Proofreading] Some parts of the sentence needed a bit of clarity, so consider using the example below to help you. 
```suggestion
In this simple animation, the background and the text of the button change depending on the value of the `isanimation` variable. When we press the button, the `isanimation` changes which activate the implicit animation and smoothly change the button's text and background (try to remove the animation to see the difference). As the name implies, SwiftUI creates and manages implicit animations whenever specific situations or user interface state changes take place, which means developers don't have full control over them.
```

## Explicit Animation using withAnimation(_:_:)

> To be able to make Explicit animation, we use the same three steps we used in the implicit animation. this example uses Explicit animation to change the position of a button:

Reviewer: [fix]: I found some typos that need fixing.
```suggestion
To be able to make Explicit animation, we use the same three steps we used in the implicit animation. The below code snippet uses Explicit animation to change the position of a button:
```

```Swift
struct ImplicitAnimation: View {
    @State var isanimation = false // @State variable that changes to activate the animation.
    
    var body: some View {
        Button {
            withAnimation(.spring()) {
                isanimation.toggle() // This is the Implicit animation.
            }
        } label: {
            Text("press me")
                .foregroundColor(.white)
                .padding(.all)
                .background(Color.black)
        }
        .offset(x: isanimation ? -20 : 60,y:0) // Ternary Operator used to change the view based onn the value of the animation.

    }
}
  ```

> **Note:** `withAnmation` isn't a view modifier like `.animation`, it's a function that needs to be called to preform the change to the state in a smooth way, which means that any view that uses the same state in there view will also be animated.

![This is a gif that shows the effect of Explicit animation on the view.](https://raw.githubusercontent.com/Codecademy/docs/main/media/Explicit-Animation-Swiftui.gif)

> As you see from the output, we used the `withAnimation` to change the poisition of the button with a spring effect to it. what is beautiful about explicit animation is that you can animate diffrent views with the same call of `withAinmation`, but the two diffrent views must use the same state variable to do that, for example, to animate the background color of the label view of the button and the position of the button at the same time a ternary operator will be added to `.background()` modifier. it looks like that:

Reviewer: ```suggestion
As you see from the output, we used the `withAnimation` to change the poisition of the button with a spring effect to it. what is beautiful about explicit animation is that you can animate different views with the same call of `withAinmation`, but the two different views must use the same state variable to do that. For example, to animate the background color of the label view of the button and the position of the button at the same time, a ternary operator will be added to `.background()` modifier. It looks like this:
```

```Swift
struct ImplicitAnimation: View {
    @State var isanimation = false // @State variable that changes to activate the animation.
    
    var body: some View {
        Button {
            withAnimation(.spring()) {
                isanimation.toggle() // This is the Implicit animation.
            }
        } label: {
            Text("press me")
                .foregroundColor(.white)
                .padding(.all)
                .background(isanimation ? Color.blue : Color.black) // Ternary Operator used to change the view based onn the value of the animation.
        }
        .offset(x: isanimation ? -20 : 60,y:0) // Ternary Operator used to change the view based onn the value of the animation.

    }
}
  ```
![This is a gif that shows the effect of Explicit animation on two diffrent views using the same state variable.](https://raw.githubusercontent.com/Codecademy/docs/main/media/morethanoneView-with-Explicit-Animation-Swiftui.gif)

> The last thing, there is no limit on how `withAnimation` or `.animation` can be used. What is explained above is just the first step towards understanding how to do animation in swiftUI, be creative with it, it's a very powerful tool in swiftUI that enhances the user experience of any app just by adding it to your app.

Reviewer: ```suggestion
One thing to consider is that no limit on how `withAnimation` or `.animation` can be used. This is a very powerful tool that can be used in many ways. 
```
